lib/chef/resource/chocolatey_installer.rb

```diff
@@ -0,0 +1,198 @@
+class Chef
+  class Resource
+    class ChocolateyInstaller < Chef::Resource
+      provides :chocolatey_installer
+
+      description "Use the Chocolatey Installer resource to ensure that Choco is installed to your specification. Use the Chocolatey Feature resource to customize your install"
+      introduced "18.3"
+      examples <<~DOC
+          **Install Chocolatey**
+
+          ```ruby
+          chocolatey_installer 'latest' do
+            action :install
+          end
+          ```
+
+          **Uninstall Chocolatey**
+
+          ```ruby
+          chocolatey_installer 'Some random verbiage' do
+            action :uninstall
+          end
+          ```
+
+          **Install Chocolatey with Parameters**
+
+          ```ruby
+          chocolatey_installer 'latest' do
+            action :install
+            download_url "https://www.contoso.com/foo"
+            chocolatey_version '2.12.24'
+          end
+          ```
+
+          **Upgrade Chocolatey with Parameters**
+
+          ```ruby
+          chocolatey_installer 'latest' do
+            action :upgrade
+            chocolatey_version '2.12.24'
+          end
+          ```
+      DOC
+
+      allowed_actions :install, :uninstall, :upgrade
+
+      property :download_url, String,
+        description: "The URL to download Chocolatey from. This defaults to the value of $env:ChocolateyDownloadUrl, if it is set, and otherwise falls back to the official Chocolatey community repository to download the Chocolatey package. It can be used for offline installation by providing a path to a Chocolatey.nupkg."
```

Comment:
We don't usually consult env vars from cookbooks.  It breaks idempotency... Two prior with different environments running this could move the install back and forth between different versions. I highly highly recommend against this.

Reply:
I have a question - this defaults to a choco repo if you specify nothing. The point of this url is the use case where a random update needs to be downloaded from an alternate source. It does not need to be idempotent.

Reply:
So, without involving the environment variable, we would have the default location and this override? Seems like having the environment variable as yet another backup configuration location might be overkill, anyway. Just default to the standard url and have this as an override.

Reply:
Per our discussion, this is actually being used directly by the Chocolatey `install.ps1` script and the environment variable can be used to redirect the standard installer to an alternate location.

Reply:
@johnmccrae can we made the description more clear that it's the choco installer itself looking at these variables?

Reply:
done and spelling error corrected too